controller-client/src/lib.rs

```diff
@@ -556,12 +556,28 @@ impl ControllerClientImpl {
     /// which runs the connection in a background task and provides a `mpsc` channel interface.
     /// Due to this cloning the `Channel` type is cheap and encouraged.
     ///
-    fn get_controller_client(&self) -> ControllerServiceClient<Channel> {
-        self.channel.read().unwrap().clone()
+    async fn get_controller_client(&self) -> ControllerServiceClient<Channel> {
+        if self.config.is_auth_enabled && self.config.credentials.is_expired() {
+            // get_request_metadata internally checks if token expired before sending request to the server,
+            // race condition might happen here but eventually only one request will be sent.
+            let mut x = self.channel.write().await;
```

Comment:
we should remove code duplication of this logic here and `reset()`.

Reply:
Thanks!

Reply:
Just realize that we cannot remove the logic in reset. The reason is that `basic` method will never expire so if we remove the logic in reset then after resetting the auth header will be gone.